etl/steps/data/garden/climate_watch/2023-10-31/emissions_by_sector.py

```diff
@@ -0,0 +1,224 @@
+# Convert million tonnes to tonnes.
+MT_TO_T = 1e6
+
+
+def add_region_aggregates(tb: Table, ds_regions: Dataset, ds_income_groups: Dataset) -> Table:
+    tb_regions = tb.copy()
+
+    # Create region aggregates for all columns (with a simple sum) except for the column of efficiency factors.
+    aggregations = {
+        column: "sum" for column in tb_regions.columns if column not in ["country", "year", "efficiency_factor"]
+    }
+
+    # Add region aggregates.
+    for region in REGIONS:
+        members = geo.list_members_of_region(
+            region=region,
+            ds_regions=ds_regions,
+            ds_income_groups=ds_income_groups,
+            additional_members=REGIONS[region].get("additional_members"),
+            include_historical_regions_in_income_groups=True,
+        )
+        tb_regions = geo.add_region_aggregates(
+            df=tb_regions,
+            region=region,
+            aggregations=aggregations,
+            countries_in_region=members,
+            countries_that_must_have_data=[],
+            num_allowed_nans_per_year=None,
+            frac_allowed_nans_per_year=0.9999,
+        )
+    # Copy metadata of original table.
+    tb_regions = tb_regions.copy_metadata(from_table=tb)
+
+    return tb_regions
+
+
+def create_table_for_gas(
+    tb: Table, gas: str, ds_regions: Dataset, ds_population: Dataset, ds_income_groups: Dataset
+) -> Table:
+    """Extract data for a particular gas and create a table with variables' metadata.
+
+    Parameters
+    ----------
+    tb : Table
+    gas : str
+        Name of gas to consider (as called in "gas" column of the original data).
+    ds_regions : Dataset
+        Regions dataset.
+    ds_population : Dataset
+        Population dataset.
+    ds_income_groups : Dataset
+        Income groups dataset.
+
+    Returns
+    -------
+    table_gas : Table
+        Table with data for considered gas, and metadata for each variable.
+
+    """
+    # Select data for current gas.
+    tb_gas = tb[tb["gas"] == gas].drop(columns="gas").reset_index(drop=True)
+
+    # Pivot table to have a column for each sector.
+    tb_gas = tb_gas.pivot(index=["country", "year"], columns="sector", values="value").reset_index()
+
+    # Create region aggregates.
+    tb_gas = add_region_aggregates(tb=tb_gas, ds_regions=ds_regions, ds_income_groups=ds_income_groups)
+
+    # Add population to data.
+    tb_gas = geo.add_population_to_table(tb=tb_gas, ds_population=ds_population)
+
+    # List columns with emissions data.
+    emissions_columns = [column for column in tb_gas.columns if column not in ["country", "year", "population"]]
+
+    # Add per capita variables.
+    for variable in emissions_columns:
+        tb_gas[variable + PER_CAPITA_SUFFIX] = MT_TO_T * tb_gas[variable] / tb_gas["population"]
+
+    # Remove rows and columns that only have nans.
+    tb_gas = tb_gas.dropna(how="all", axis=1)
+    tb_gas = tb_gas.dropna(subset=emissions_columns, how="all").reset_index(drop=True)
```

Comment:
Shouldn't this be before adding per capita variables? Some per capita columns could stay empty after the calculation

Reply:
It doesn't make a difference (in fact there are no rows or columns with nans.